service/qos/service_level_controller.cc

```diff
@@ -220,6 +223,64 @@ future<> service_level_controller::update_service_levels_cache() {
     });
 }
 
+future<> service_level_controller::update_effective_service_levels_cache() {
+    assert(this_shard_id() == global_controller);
+    
+    if (!_auth_service.local_is_initialized()) {
+        // Because cache update is triggered in `topology_state_load()`, auth service
+        // might be not initialized yet.
+        return make_ready_future<>();
+    }
+
+    return with_semaphore(_global_controller_db->notifications_serializer, 1, [this] () -> future<> {
+        auto& role_manager = _auth_service.local().underlying_role_manager();
+        const auto all_roles = co_await role_manager.query_all();
+        const auto hierarchy = co_await role_manager.query_all_directly_granted();
+        // includes only roles with attached service level
+        const auto attributes = role_manager.query_attribute_for_all("service_level").get();
+
+        std::map<sstring, service_level_options> effective_sl_map;
+
+        auto sorted = co_await utils::topological_sort(all_roles, hierarchy);
+        // Roles are sorted from the top of the hierarchy to the botttom. 
+        /// `GRANT role1 TO role2` means role2 is higher in the hierarchy than role1, so role2 will be before
+        // role1 in `sorted` vector.
+        // That's why if we iterate over the vector in reversed order, we will visit the roles from the bottom
+        // and we can use already calculated effective service levels for all of the subroles.
+        for (auto& role: sorted | boost::adaptors::reversed) {
+            std::optional<service_level_options> sl_options;
+
+            if (auto sl_name_it = attributes.find(role); sl_name_it != attributes.end()) {
+                auto sl = _service_levels_db.at(sl_name_it->second);
+                sl_options = sl.slo;
+                sl_options->init_effective_names(sl_name_it->second);
+            }
+
+            auto [it, it_end] = hierarchy.equal_range(role);
+            while (it != it_end) {
+                auto& subrole = it->second;
+                if (effective_sl_map.contains(subrole)) {
+                    auto& sub_sl = effective_sl_map.at(subrole);
```

Comment:
Nit: why not do `if (auto effective_sl_it = effective_sl_map.find(subrole); effective_sl_it != effective_sl_map.end()) {}` and avoid double lookup?